### PR TITLE
API: Provide PubMed IDs rather than raw text

### PR DESCRIPTION
**Pull request by jvwong**
Add the ability to retrieve article text if PMID is provided. Description in the updated README should suffice (below)

Notes:
  - approach
    - uses a validator to map input to the target `Query` model 
    - backwards compatible
  - environment variables
    - setting and retrieving things like app name, version. I just ended up duplicating this in the .env file and reading them off
  - performance
    - fetching from pubmed takes ~0.015s/uid and scales linearly.
    - e.g. for 12 177 uids it took around 3.5 minutes

Refs #10

--- 

## Usage

...

Once the server is running, you can make a POST request with:

1. JSON body that is self-contained. Provide the text in `query` and text in `documents` to search against. Sample JSON request:

    ```json
    {
        "query": {
            "uid":"9887103",
            "text": "The Drosophila activin receptor baboon signals through dSmad2 and controls cell proliferation but not patterning during larval development. The TGF-beta superfamily of growth and differentiation factors, including TGF-beta, Activins and bone morphogenetic proteins (BMPs) play critical roles in regulating the development of many organisms. These factors signal through a heteromeric complex of type I and II serine/threonine kinase receptors that phosphorylate members of the Smad family of transcription factors, thereby promoting their nuclear localization. Although components of TGF-beta/Activin signaling pathways are well defined in vertebrates, no such pathway has been clearly defined in invertebrates. In this study we describe the role of Baboon (Babo), a type I Activin receptor previously called Atr-I, in Drosophila development and characterize aspects of the Babo intracellular signal-transduction pathway. Genetic analysis of babo loss-of-function mutants and ectopic activation studies indicate that Babo signaling plays a role in regulating cell proliferation. In mammalian cells, activated Babo specifically stimulates Smad2-dependent pathways to induce TGF-beta/Activin-responsive promoters but not BMP-responsive elements. Furthermore, we identify a new Drosophila Smad, termed dSmad2, that is most closely related to vertebrate Smads 2 and 3. Activated Babo associates with dSmad2 but not Mad, phosphorylates the carboxy-terminal SSXS motif and induces heteromeric complex formation with Medea, the Drosophila Smad4 homolog. Our results define a novel Drosophila Activin/TGF-beta pathway that is analogous to its vertebrate counterpart and show that this pathway functions to promote cellular growth with minimal effects on patterning."
        },
        "documents":[
            {
                "uid": "9887103",
                "text": "The Drosophila activin receptor baboon signals through dSmad2 and controls cell proliferation but not patterning during larval development. The TGF-beta superfamily of growth and differentiation factors, including TGF-beta, Activins and bone morphogenetic proteins (BMPs) play critical roles in regulating the development of many organisms. These factors signal through a heteromeric complex of type I and II serine/threonine kinase receptors that phosphorylate members of the Smad family of transcription factors, thereby promoting their nuclear localization. Although components of TGF-beta/Activin signaling pathways are well defined in vertebrates, no such pathway has been clearly defined in invertebrates. In this study we describe the role of Baboon (Babo), a type I Activin receptor previously called Atr-I, in Drosophila development and characterize aspects of the Babo intracellular signal-transduction pathway. Genetic analysis of babo loss-of-function mutants and ectopic activation studies indicate that Babo signaling plays a role in regulating cell proliferation. In mammalian cells, activated Babo specifically stimulates Smad2-dependent pathways to induce TGF-beta/Activin-responsive promoters but not BMP-responsive elements. Furthermore, we identify a new Drosophila Smad, termed dSmad2, that is most closely related to vertebrate Smads 2 and 3. Activated Babo associates with dSmad2 but not Mad, phosphorylates the carboxy-terminal SSXS motif and induces heteromeric complex formation with Medea, the Drosophila Smad4 homolog. Our results define a novel Drosophila Activin/TGF-beta pathway that is analogous to its vertebrate counterpart and show that this pathway functions to promote cellular growth with minimal effects on patterning."
            },
            {
                "uid": "30049242",
                "text": "Transcriptional up-regulation of the TGF-β intracellular signaling transducer Mad of Drosophila larvae in response to parasitic nematode infection. The common fruit fly Drosophila melanogaster is an exceptional model for dissecting innate immunity. However, our knowledge on responses to parasitic nematode infections still lags behind. Recent studies have demonstrated that the well-conserved TGF-β signaling pathway participates in immune processes of the fly, including the anti-nematode response. To elucidate the molecular basis of TGF-β anti-nematode activity, we performed a transcript level analysis of different TGF-β signaling components following infection of D. melanogaster larvae with the nematode parasite Heterorhabditis gerrardi. We found no significant changes in the transcript level of most extracellular ligands in both bone morphogenic protein (BMP) and activin branches of the TGF-β signaling pathway between nematode-infected larvae and uninfected controls. However, extracellular ligand, Scw, and Type I receptor, Sax, in the BMP pathway as well as the Type I receptor, Babo, in the activin pathway were substantially up-regulated following H. gerrardi infection. Our results suggest that receptor up-regulation leads to transcriptional up-regulation of the intracellular component Mad in response to H. gerrardi following changes in gene expression of intracellular receptors of both TGF-β signaling branches. These findings identify the involvement of certain TGF-β signaling pathway components in the immune signal transduction of D. melanogaster larvae against parasitic nematodes ."
            },
            {
                "uid": "22936248",
                "text": "High-fidelity promoter profiling reveals widespread alternative promoter usage and transposon-driven developmental gene expression. Many eukaryotic genes possess multiple alternative promoters with distinct expression specificities. Therefore, comprehensively annotating promoters and deciphering their individual regulatory dynamics is critical for gene expression profiling applications and for our understanding of regulatory complexity. We introduce RAMPAGE, a novel promoter activity profiling approach that combines extremely specific 5'-complete cDNA sequencing with an integrated data analysis workflow, to address the limitations of current techniques. RAMPAGE features a streamlined protocol for fast and easy generation of highly multiplexed sequencing libraries, offers very high transcription start site specificity, generates accurate and reproducible promoter expression measurements, and yields extensive transcript connectivity information through paired-end cDNA sequencing. We used RAMPAGE in a genome-wide study of promoter activity throughout 36 stages of the life cycle of Drosophila melanogaster, and describe here a comprehensive data set that represents the first available developmental time-course of promoter usage. We found that >40% of developmentally expressed genes have at least two promoters and that alternative promoters generally implement distinct regulatory programs. Transposable elements, long proposed to play a central role in the evolution of their host genomes through their ability to regulate gene expression, contribute at least 1300 promoters shaping the developmental transcriptome of D. melanogaster. Hundreds of these promoters drive the expression of annotated genes, and transposons often impart their own expression specificity upon the genes they regulate. These observations provide support for the theory that transposons may drive regulatory innovation through the distribution of stereotyped cis-regulatory modules throughout their host genomes."
            }
        ],
        "top_k":3
    }
    ```

    The return value is a JSON representation of the `top_k` most similar documents (default: return all):

    ```json
    [
        {
            "uid": "9887103",
            "score": 1.0
        },
        {
            "uid": "30049242",
            "score": 0.6427373886108398
        },
        {
            "uid": "22936248",
            "score": 0.49102723598480225
        }
    ]
    ```

    - NB: In this case, each `uid` in `documents` should be unique, but otherwise have no meaning.

2. JSON body that references PubMed article uids. Sample JSON request:

    ```json
    {
        "query": "9887103",
        "documents":["9887103", "30049242", "22936248"],
        "top_k":3
    }
    ```

    - Notes:
      - For each Document element, the text consists of the `ArticleTitle` appended to `Abstract` for that PubMed article. See [pubmed DTD](https://dtd.nlm.nih.gov/ncbi/pubmed/doc/out/180101/index.html)
      - JSON body may consist of either objects (as in Case 1) or PMID strings for `query` and elements of `documents`. However, the elements of `documents` must either be all be a single type.

### CODE REVIEW

**Comment by maxkfranz** [comment]
Looks good. 

Is it possible to do the eutils queries in batch or in parallel?  The duration probably scales linearly because of all the individual requests. 

Another thing to consider is that we’d have to make duplicate queries in order to get full article metadata on the Biofactoid side. So at this point, I’m not certain whether we’d see a significant performance difference. It’s a great first step, in any case.

File: README.md [hunk 1]
```diff
@@ -43,11 +43,68 @@ To provide arguments to the server, pass them as environment variables, e.g.:
...
...start site specificity, generates accurate and reproducible promoter expression measurements, and yields extensive transcript connectivity information through paired-end cDNA sequencing. We used RAMPAGE in a genome-wide study of promoter activity throughout 36 stages of the life cycle of Drosophila melanogaster, and describe here a comprehensive data set that represents the first available developmental time-course of promoter usage. We found that >40% of developmentally expressed genes have at least two promoters and that alternative promoters generally implement distinct regulatory programs. Transposable elements, long proposed to play a central role in the evolution of their host genomes through their ability to regulate gene expression, contribute at least 1300 promoters shaping the developmental transcriptome of D. melanogaster. Hundreds of these promoters drive the expression of annotated genes, and transposons often impart their own expression specificity upon the genes they regulate. These observations provide support for the theory that transposons may drive regulatory innovation through the distribution of stereotyped cis-regulatory modules throughout their host genomes."
```
Comment (JohnGiorgi) [inline README.md:67]: Minor comment but what if the readme examples were much shorter (e.g. single sentences?). I know this is for internal use but the readme would be a little easier to digest this way

**Comment by jvwong** [comment]
> Is it possible to do the eutils queries in batch or in parallel? The duration probably scales linearly because of all the individual requests in series.

The loop is doing querying EUTILS 10000 PMIDs at a time. That loop is there if you have more than 10000 ids. 
> 
> Another thing to consider is that we’d have to make duplicate queries in order to get full article metadata on the Biofactoid side. So at this point, I’m not certain whether we’d see a significant performance difference. It’s a great first step, in any case.

My suggestion is that Factoid do its own fetch for the few (i.e. ~30) articles it will store/show in the UI --  at the very end. Otherwise, Factoid should only be passing around lists of uids (ideally).

**Comment by jvwong** [comment]
> Now that we have two modules `ncbi.py` and `main.py` we should probably refactor this as a package. This would mean adding a directory (`semantic_search`) under the top-level of the source tree and putting `ncbi.py` and `main.py` under that folder. I can do this if there are no complaints.

Sure if you can do this that's great. I have no experience with these best practice type things. 
Also, do you know how we can set/get things like 'version' and 'name' etc in setup.py? I would use them for identifying info in requests.

**Comment by JohnGiorgi** [comment]
> Sure if you can do this that's great. I have no experience with these best practice type things.

Okay, I did the simple refactor. In the future any new modules can go under `semantic_search/my_new_module.py`. I checked that all the examples in the README work with this new code and added a missing dependency to `setup.py` that was missing (`xmltodict`). I also formatted and linted the codebase.

> Also, do you know how we can set/get things like 'version' and 'name' etc in setup.py? I would use them for identifying info in requests.

Good question. There may be advice somewhere in these docs: https://fastapi.tiangolo.com/ to get that information in requests.

From a package standpoint, I actually don't think there is a standard in python. After some googling, I am seeing `pbr` being suggested a bunch: https://docs.openstack.org/pbr/latest/. Maybe we can update this to be a `pbr` managed package in a seperate PR?

**Comment by jvwong** [comment]
> > Sure if you can do this that's great. I have no experience with these best practice type things.
> 
> Okay, I did the simple refactor. In the future any new modules can go under `semantic_search/my_new_module.py`. I checked that all the examples in the README work with this new code and added a dependency to `setup.py` that was missing (`xmltodict`). I also formatted and linted the codebase.

Good catch. 
 
> > Also, do you know how we can set/get things like 'version' and 'name' etc in setup.py? I would use them for identifying info in requests.
> 
> Good question. There may be advice somewhere in these docs: https://fastapi.tiangolo.com/ to get that information in requests.
> 
> From a package standpoint, I actually don't think there is a standard in python. After some googling, I am seeing `pbr` being suggested a bunch: https://docs.openstack.org/pbr/latest/. Maybe we can update this to be a `pbr` managed package in a seperate PR?

Yeah lemme look around at that. Right now I'm copying it into .env. Its not a big deal right now.

**Comment by maxkfranz** [comment]
> The loop is querying EUTILS with 10000 PMIDs at a time. That loop is there if you have more than 10000 ids.

I didn't notice that.  Thanks for letting me know

> My suggestion is that Factoid do its own fetch for the few (i.e. ~30) articles it will store/show in the UI -- at the very end. Otherwise, Factoid should only be passing around lists of uids (ideally).

Sounds good

**Comment by jvwong** [comment]
@JohnGiorgi this works for me - are you thumbs up?

**Comment by JohnGiorgi** [comment]
> @JohnGiorgi this works for me - are you thumbs up?

Yup feel free to merge

**Pull request merged**